### DISCUSSION TITLE
Make swift operate on its own

### DISCUSSION THREAD

**Discussion by BECATRUE** [snu-quiqcl/qiwis#146] [Ideas]
Now, to start swift, first of all we should make an instance of `QApplication`.
But, it has two main problems.

1. Intefere with the independence of swift
Once we want to start swift, we should make `QApplication`.
I think moving the part into swift is more convenient and makes swift independent.

2. Cannot add a new function in the constructor of swift
When we develop ARTIQ dashboard later, we should connect with ARTIQ system before swift operates.
For this, adding this part into the constructor of swift looks good, but in the current structure, we cannot modify the constructor.

Thus, I suggest as follows.
```python
swift = Swift(setup_path)  # only the path of setup.json
```
It means that all procedure of current `main` function will be moved in the constuctor of swift, such as parsing setup.json and create a `QApplication`.

**Comment by BECATRUE** [comment]
Please leave your opinion, freely! @kangz12345

**Comment by kangz12345** [comment]
Thank you for elaborating on this topic!
My opinions:
1. I don't think that is a problem. The PyQt system requires it, so we should follow it. If we want to instantiate a widget and show it, we will do the same thing, i.e., create a `QApplication` object first, and then instantiate the widget and `show()` it.
2. Could you please explain more details about what should we do for ARTIQ widgets? I thought separating the procedures gives less dependence. I don't see how are you going to fix which problems by changing the `Swift` constructor.

**Reply by BECATRUE** [reply]
Umm.. First of all, I modified my goal as __to allow you to customize the constructor of swift__.

To set up ARTIQ, we should operate as follows (just an example):
```python
def main():
    args = get_argparser().parse_args()
    app_infos = _read_setup_file(args.setup_path)

    qapp = QApplication(sys.argv)
    swift = Swift(app_infos)
    loop = QEventLoop(qapp)
    asyncio.set_event_loop(loop)
    atexit.register(loop.close)

    # create connections to master
    client = AsyncioClient()
    loop.run_until_complete(client.connect_rpc(
        args.server, args.port_control, "master_schedule"))
    atexit.register(client.close_rpc)
    rpc_clients = {"schedule": client}

    Client(args.server, args.port_control, "master_config")

    explist = {
        'Simple simulation': {
            'file': 'simple_simulation.py',
            'class_name': 'SimpleSimulation',
            'arginfo': dict(),
            'argument_ui': None,
            'scheduler_defaults': {}
        }
    }
    expmgr = experiments.ExperimentManager(swift.mainWindow, rpc_clients["schedule"], explist)

    qapp.exec_()
```
This process must be done before `swift.load()` to operate ARTIQ well, so we can modify the constructor of swift freely.

**Reply by BECATRUE** [reply]
So, I suggest two ideas.

1. Just make a subclass of swift.
```python
class ArtiqSwift(Swift):
  def __init__(self, ...):
    # operate desired operation
    super().__init__()
```

2. Like apps, make external module and add it into 'setup.json'.

setup.json
```json
{
  "constructor": {
    ...
  },
  "app": {
    "app1": {},
    "app2": {},
    ...
  }
}
```
swift,py
```python
class Swift():
  def __init__(self, ...):
    ...
    # read constructor file from setup.json and operate it
    self.load(appInfo)
    ...
```

**Reply by kangz12345** [reply]
> Umm.. First of all, I modified my goal as **to allow you to customize the constructor of swift**.
> 
> To set up ARTIQ, we should operate as follows (just an example):
> 
> ```python
> def main():
>     args = get_argparser().parse_args()
>     app_infos = _read_setup_file(args.setup_path)
> 
>     qapp = QApplication(sys.argv)
>     swift = Swift(app_infos)
>     loop = QEventLoop(qapp)
>     asyncio.set_event_loop(loop)
>     atexit.register(loop.close)
> 
>     # create connections to master
>     client = AsyncioClient()
>     loop.run_until_complete(client.connect_rpc(
>         args.server, args.port_control, "master_schedule"))
>     atexit.register(client.close_rpc)
>     rpc_clients = {"schedule": client}
> 
>     Client(args.server, args.port_control, "master_config")
> 
>     explist = {
>         'Simple simulation': {
>             'file': 'simple_simulation.py',
>             'class_name': 'SimpleSimulation',
>             'arginfo': dict(),
>             'argument_ui': None,
>             'scheduler_defaults': {}
>         }
>     }
>     expmgr = experiments.ExperimentManager(swift.mainWindow, rpc_clients["schedule"], explist)
> 
>     qapp.exec_()
> ```
> 
> This process must be done before `swift.load()` to operate ARTIQ well, so we can modify the constructor of swift freely.

So this code is actually not okay because it executes `load()` when we construct a `Swift`?

**Reply by kangz12345** [reply]
After we execute the codes above, does the apps (e.g., ARTIQ scheduler, controller, experiment builder, etc.) need the reference of any variables, e.g., `client`, `loop`, `explist`, `expmgr`, etc.?
In other words, are the codes above entirely independent to the apps?

**Reply by BECATRUE** [reply]
No! I haven't seen all the functions such as schedulers yet, so I'm not sure exactly what it needs, but _it needs at least `expmgr`.
Furthermore, the gui shows the directory structure where `artiq_master` is running, more elements are necessary in `swift`.

For reference, ARTIQ also placed these elements within the gui!

**Reply by BECATRUE** [reply]
However, I'm a little worried about the ARTIQ structure is too complex..
